File: src/main/java/name/seguri/java/leetcode/_21/ListNode.java
package name.seguri.java.leetcode._21;

import java.util.StringJoiner;

/** An empty list is represented by a null node. */
class ListNode {
  int val;
  ListNode next;

  ListNode() {}

  ListNode(int val) {
    this.val = val;
  }

  ListNode(int val, ListNode next) {
    this.val = val;
    this.next = next;
  }

  /**
   * Build the nodes from the tail. If you start from the head, you need to keep two references, one
   * for the head and one for the current node.
   */
  public static ListNode of(int... values) {
    ListNode curr = null;
    for (int i = values.length - 1; i >= 0; i--) {
      curr = new ListNode(values[i], curr);
    }
    return curr;
  }

  @Override
  public String toString() {
    var sj = new StringJoiner(", ", "[", "]");
    var curr = this;
    while (curr != null) {
      sj.add(String.valueOf(curr.val));
      curr = curr.next;
    }
    return sj.toString();
  }
}

File: src/main/kotlin/name/seguri/kotlin/leetcode/_21/ListNode.kt
package name.seguri.kotlin.leetcode._21

import java.util.StringJoiner

/** I've preferred this over `class ListNode(var `val`: Int, var next: ListNode?)` */
class ListNode(var `val`: Int) {
  var next: ListNode? = null

  companion object {
    fun of(vararg values: Int): ListNode? {
      var curr: ListNode? = null
      for (i in values.indices.reversed()) {
        curr = ListNode(values[i]).apply { next = curr }
      }
      return curr
    }
  }

  override fun toString(): String {
    val sj = StringJoiner(", ", "[", "]")
    var curr: ListNode? = this
    while (curr != null) {
      sj.add(curr.`val`.toString())
      curr = curr.next
    }
    return sj.toString()
  }
}
